File: src-tauri/src/commands/genui.rs
use crate::models::AppError;
use crate::services::GenUiService;
use crate::storage::types::{CreateGenUiRequest, GenUi, UpdateGenUiRequest, UUID};
use tauri::State;

#[tauri::command]
pub async fn genui_create(
    request: CreateGenUiRequest,
    genui_service: State<'_, GenUiService>,
) -> Result<GenUi, AppError> {
    genui_service.create_genui(request.name, request.spec).await
}

#[tauri::command]
pub async fn genui_list(
    limit: Option<i32>,
    offset: Option<i32>,
    genui_service: State<'_, GenUiService>,
) -> Result<Vec<GenUi>, AppError> {
    genui_service.list_genui(limit, offset).await
}

#[tauri::command]
pub async fn genui_get(
    genui_id: UUID,
    genui_service: State<'_, GenUiService>,
) -> Result<GenUi, AppError> {
    genui_service.get_genui(genui_id).await
}

#[tauri::command]
pub async fn genui_update(
    genui_id: UUID,
    request: UpdateGenUiRequest,
    genui_service: State<'_, GenUiService>,
) -> Result<GenUi, AppError> {
    genui_service
        .update_genui(genui_id, request.name, request.spec)
        .await
}

#[tauri::command]
pub async fn genui_delete(
    genui_id: UUID,
    genui_service: State<'_, GenUiService>,
) -> Result<(), AppError> {
    genui_service.delete_genui(genui_id).await
}

File: src/lib/api/genui.ts
/**
 * A GenUI is a named, reusable JSON-Render UI spec. The backend stores `spec`
 * as opaque text; callers validate it with explainSpec before saving.
 */

import { apiCall } from "./index";
import type { GenUi, UUID } from "../types";

export async function createGenui(name: string, spec: string): Promise<GenUi> {
  return apiCall<GenUi>("genui_create", { request: { name, spec } });
}

/** Ordered by updatedAt descending. */
export async function getGenuis(
  limit?: number,
  offset?: number,
): Promise<GenUi[]> {
  return apiCall<GenUi[]>("genui_list", { limit, offset });
}

export async function getGenui(genuiId: UUID): Promise<GenUi> {
  return apiCall<GenUi>("genui_get", { genuiId });
}

/** Omitted fields are left unchanged. */
export async function updateGenui(
  genuiId: UUID,
  name?: string,
  spec?: string,
): Promise<GenUi> {
  return apiCall<GenUi>("genui_update", {
    genuiId,
    request: { name, spec },
  });
}

/** The backend also clears any agent.genuiId referencing it. */
export async function deleteGenui(genuiId: UUID): Promise<void> {
  return apiCall<void>("genui_delete", { genuiId });
}
